doc/pj_validate.1.txt
PJ_VALIDATE(1)
==============
:doctype: manpage


NAME
----
pj_validate - checks a file strictly follows the Paje file format


SYNOPSIS
--------
*pj_validate* ['OPTIONS'] ['FILE']


DESCRIPTION
-----------

The pj_validate(1) command checks if the paje trace file 'FILE'
strictly follows the Paje file format as defined by the reference
documentation (see section RESOURCES for the appropriate link). It is
a useful program to check if a Paje file is valid and can be used in
the performance analysis of parallel and distributed applications.  If
'FILE' is not provided, then the standard input is used.

*pj_validate* reads the trace file from the beginning until the end of
file is found. During this process, *pj_validate* relies on the Paje
library to recreate in memory the behavior that is registered in the
trace file. This means that *pj_validate* will put in memory all the
contents of the trace file, even if the input is very large. Once all
the contents of the trace file are simulated without any kind of
error, *pj_validate* exits with a value of 0 (meaning the everything
is okay). If the trace is valid, the *pj_validate* also outputs
information regarding the trace: the total number of containers in the
trace, the total number of entities (states, events, links and
variables) and the type hierarchy contained in the trace.


OPTIONS
-------

*pj_validate* accepts the following options:    

*-n, --no-strict*::
    Support old field names in event definitions.

*-f, --flex*::
    Use alternative file reader based on flex/bison (experimental).

*-q, --quiet*::
    Be quiet (no output on stdout).

*-t, --time*::
    Print number of seconds to simulate input

*-?, --help*::
    Show all the available options.

*--usage*::
    Give a short usage message.

INPUT DESCRIPTION
-----------------

The pj_validate(1) command expects an input that follows the Paje file
format (as described in the PDF document listed in the RESOURCES
section of this page). If FILE is not provided, pj_validate(1) will try
to read from the standard input.

OUTPUT DESCRIPTION
------------------

If the input strictly follows the Paje file format, the output of the
pj_validate(1) command looks like this:

  report Containers: 82
  report Entities: 11417509
  report Types: 6
  report Type Hierarchy:
  report |0 (17PajeContainerType)
  report ||ROOT (17PajeContainerType)
  report || LINK (12PajeLinkType)
  report || MACHINE (17PajeContainerType)
  report | | PROCESS (17PajeContainerType)
  report |  | STATE (13PajeStateType)

for an input that contains 82 containers, 11417509 entities spread
among all the containers and 6 types. The type hierarchy starts with
the default root identified by 0 (a container type), with one child
named ROOT (also a container type) that has two children: LINK (a link
type) and MACHINE (a container type). MACHINE has one child PROCESS (a
container type) that has a single child which is a STATE (state type).

RESOURCES
---------

Description of the Paje trace file:
<http://paje.sourceforge.net/download/publication/lang-paje.pdf>

Main web site:
<http://github.com/schnorr/pajeng/>

REPORTING BUGS
--------------

       Report pj_validate bugs to <http://github.com/schnorr/pajeng/issues>


COPYRIGHT
---------

Copyright \(C) 2012-2014 Lucas M. Schnorr. Free use of this software is granted under the terms of the GNU General Public License (GPL).


SEE ALSO
--------

*pj_dump(1)*


src/pj_validate/pj_validate.cc
/*
    This file is part of PajeNG

    PajeNG is free software: you can redistribute it and/or modify
    it under the terms of the GNU Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    PajeNG is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Public License for more details.

    You should have received a copy of the GNU Public License
    along with PajeNG. If not, see <http://www.gnu.org/licenses/>.
*/
#include <sys/time.h>
#include <string>
#include <iostream>
#include <exception>
#include "PajeFileReader.h"
#include "PajeFlexReader.h"
#include "PajeEventDecoder.h"
#include "PajeSimulator.h"
#include "PajeException.h"
#include <argp.h>

#define VALIDATE_INPUT_SIZE 2
static char doc[] = "Checks if FILE, or standard input, strictly follows the Paje file format definition";
static char args_doc[] = "[FILE]";

static struct argp_option options[] = {
  {"no-strict", 'n', 0, OPTION_ARG_OPTIONAL, "Support old field names in event definitions"},
  {"quiet", 'q', 0, OPTION_ARG_OPTIONAL, "Be quiet"},
  {"time", 't', 0, OPTION_ARG_OPTIONAL, "Print number of seconds to simulate input"},
  {"flex", 'f', 0, OPTION_ARG_OPTIONAL, "Use flex-based file reader"},
  { 0 }
};

struct arguments {
  char *input[VALIDATE_INPUT_SIZE];
  int noStrict;
  int input_size;
  int quiet;
  int time;
  int flex;
};

static int parse_options (int key, char *arg, struct argp_state *state)
{
  struct arguments *arguments = (struct arguments*)(state->input);
  switch (key){
  case 'n': arguments->noStrict = 1; break;
  case 't': arguments->time = 1; break;
  case 'q': arguments->quiet = 1; break;
  case 'f': arguments->flex = 1; break;
  case ARGP_KEY_ARG:
    if (arguments->input_size == VALIDATE_INPUT_SIZE) {
      /* Too many arguments. */
      argp_usage (state);
    }
    arguments->input[state->arg_num] = arg;
    arguments->input_size++;
    break;
  case ARGP_KEY_END:
    if (state->arg_num < 0) {
      /* Not enough arguments. */
      argp_usage (state);
    }
    break;
  default: return ARGP_ERR_UNKNOWN;
  }
  return 0;
}

static struct argp argp = { options, parse_options, args_doc, doc };

static double gettime (void)
{
  struct timeval tr;
  gettimeofday(&tr, NULL);
  return (double)tr.tv_sec+(double)tr.tv_usec/1000000;
}

int main (int argc, char **argv)
{
  struct arguments arguments;
  bzero (&arguments, sizeof(struct arguments));
  if (argp_parse (&argp, argc, argv, 0, 0, &arguments) == ARGP_KEY_ERROR){
    fprintf(stderr, "%s, error during the parsing of parameters\n", argv[0]);
    return 1;
  }

  PajeComponent *reader = NULL;
  PajeEventDecoder *decoder = NULL;
  PajeSimulator *simulator = NULL;

  //the global PajeDefinitions object
  PajeDefinitions *definitions = new PajeDefinitions (arguments.noStrict ? false : true); 

  try {
    //alloc reader
    if (arguments.flex){
      if (arguments.input_size == 0){
	reader = new PajeFlexReader(definitions);
      }else{
	reader = new PajeFlexReader(std::string(arguments.input[0]), definitions);
      }
    }else{
      if (arguments.input_size == 0){
	reader = new PajeFileReader();
      }else{
	reader = new PajeFileReader (std::string(arguments.input[0]));
      }
    }

    //alloc decoder and simulator
    if (!arguments.flex){
      decoder = new PajeEventDecoder(definitions);
    }
    simulator = new PajeSimulator ();

    //connect components
    if (arguments.flex){
      reader->setOutputComponent (simulator);
      simulator->setInputComponent (reader);
    }else{
      reader->setOutputComponent (decoder);
      decoder->setInputComponent (reader);
      decoder->setOutputComponent (simulator);
      simulator->setInputComponent (decoder);
    }
  }catch (PajeException& e){
    e.reportAndExit ();
  }

  //read and simulate
  double t1, t2;
  if (arguments.time){
    t1 = gettime();
  }
  try {
    reader->startReading();
    while (reader->hasMoreData()){
      reader->readNextChunk();
    }
    reader->finishedReading();
  }catch (PajeException& e){
    e.reportAndExit();
  }
  if (arguments.time){
    t2 = gettime();
    printf ("%f\n", t2-t1);
  }

  if (!arguments.quiet){
    simulator->report();
  }

  delete reader;
  delete decoder;
  delete simulator;
  return 0;
}
